### PR TITLE
Make FadeTransition more efficient

### PR DESCRIPTION
**Pull request by Hixie**
This improves the CPU side. There's also work to be done on the GPU side.

### CODE REVIEW

**Comment by Hixie** [comment]
cc @gspencergoog for review

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 1]
```diff
@@ -776,6 +777,99 @@ class RenderOpacity extends RenderProxyBox {
...
+
+  @override
+  void detach() {
+    _opacity.removeListener(_updateOpacity);
+    super.detach();
+  }
+
+  void _updateOpacity() {
+    final int oldAlpha = _alpha;
+    final bool didNeedCompositing = _currentlyNeedsCompositing;
```
Comment (gspencergoog) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: You can move this inside the if below.

File: packages/flutter/lib/src/rendering/proxy_box.dart [hunk 2]
```diff
@@ -776,6 +777,99 @@ class RenderOpacity extends RenderProxyBox {
...
+        return;
+      }
+      assert(needsCompositing);
+      context.pushOpacity(offset, _alpha, super.paint);
+    }
+  }
+
+  @override
+  void visitChildrenForSemantics(RenderObjectVisitor visitor) {
+    if (child != null && _alpha != 0)
```
Comment (gspencergoog) [inline packages/flutter/lib/src/rendering/proxy_box.dart]: Could there be times when we have an alpha of zero, but still want to visit the render object? I'm thinking about Semantics or other non-visual things (e.g. objects used for place-holding in a layout).  I can see the argument for "well, if it's not visible, then there shouldn't be any kind of representation", however.

File: packages/flutter/test/widgets/fade_transition_test.dart [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    expect(log.last, 'buildScope finished');
+    await tester.pump();
+    expect(log, hasLength(2));
+    controller.forward();
+    await tester.pumpAndSettle();
+    expect(log, hasLength(2));
+    debugPrint = oldPrint;
+    debugPrintBuildScope = false;
+  });
+}
```
Comment (gspencergoog) [inline packages/flutter/test/widgets/fade_transition_test.dart]: Newline at EOF?

File: packages/flutter/lib/src/rendering/proxy_box.dart (hunk 2 above)
Comment (Hixie) [reply packages/flutter/lib/src/rendering/proxy_box.dart]: See https://github.com/flutter/flutter/issues/12283 for the history here. This is just duplicating what Opacity does.

**Comment by Hixie** [comment]
Updated per comments. Will land on green.

**Comment by Hixie** [comment]
Wait, you didn't actually give an LGTM, never mind. :-)

PTAL! :-D

**Review by gspencergoog — APPROVED** [review]

**Pull request merged**